one-ui/src/main/java/com/marklogic/hub/oneui/auth/AuthenticationFilter.java

```diff
@@ -0,0 +1,171 @@
+ */
+public class AuthenticationFilter extends AbstractAuthenticationProcessingFilter {
+
+    private EnvironmentService environmentService;
+    private HubConfigSession hubConfig;
+
+    public AuthenticationFilter(EnvironmentService environmentService, HubConfigSession hubConfig) {
+        super(new AntPathRequestMatcher("/api/login", "POST"));
+        this.environmentService = environmentService;
+        this.hubConfig = hubConfig;
+        setAuthenticationSuccessHandler(new LoginHandler());
+        setAuthenticationFailureHandler(new LoginFailureHandler());
+    }
+
+    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException, IOException {
+        if (!request.getMethod().toUpperCase().equals("POST")) {
+            throw new AuthenticationServiceException("Authentication method not supported: " + request.getMethod());
+        }
+
+        final EnvironmentInfo originalEnvironmentInfo = environmentService.getEnvironment();
+
+        final LoginInfo loginInfo = new ObjectMapper().readValue(request.getInputStream(), LoginInfo.class);
+
+        final boolean datahubIsInstalled = StringUtils.isEmpty(loginInfo.mlHost);
+        if (!datahubIsInstalled) {
+            environmentService.setEnvironment(new EnvironmentInfo(loginInfo.mlHost, loginInfo.mlAuthMethod,
+                loginInfo.mlPort, loginInfo.mlManageAuthMethod, loginInfo.mlManagePort, loginInfo.dhStagingAuthMethod, loginInfo.dhStagingPort, loginInfo.dhFinalAuthMethod, loginInfo.dhFinalPort));
+        }
+
+        try {
+            AuthenticationToken token = authenticateUser(loginInfo.username, loginInfo.password);
+            token.setDetails(authenticationDetailsSource.buildDetails(request));
+            return token;
+        } catch (Exception e) {
+            environmentService.setEnvironment(originalEnvironmentInfo);
+            throw e;
+        }
+    }
+
+    /**
+     * @param username
+     * @param password
+     */
+    protected AuthenticationToken authenticateUser(String username, String password) {
+        if (StringUtils.isEmpty(username) || StringUtils.isEmpty(password)) {
+            throw new BadCredentialsException("Unauthorized");
+        }
+
+        username = username.trim();
+
+        EnvironmentInfo environmentInfo = environmentService.getEnvironment();
+        hubConfig.setCredentials(environmentInfo, username, password);
+
+        final boolean hasManagePrivileges = canAccessManageServer(environmentInfo.mlHost);
+
+        final DatabaseClient stagingClient = hubConfig.newStagingClient(null);
+
+        boolean dataHubInstalled = false;
+        List<GrantedAuthority> authorities = new ArrayList<>();
+        ArrayNode roles = null;
+
+        try {
+            dataHubInstalled = stagingClient.checkConnection().isConnected();
+            if (dataHubInstalled) {
+                roles = (ArrayNode) RolesService.on(stagingClient).getRoles();
```

Comment:
very nice refactors... One question is that why we need to have two services (RolesService and SecurityServicce)? Would it better use a SecurityService which consolidate the roles and authorities along with one DS call?

Reply:
BTW, I still tried to figout out @Secured, @PreAuthorize, @PostAuthorize. Another question is that do we really need to return roles of the current users to front-end. I thought that we wanted to client-side has little knowledge about roles rather than authorities (canInstallDataHub, canRead{artifactType}, canWrite{artifact})  ?

Reply:
I didn't want to get rid of RolesService and "roles" until the GUI is updated to depend on authorities instead of roles.

Reply:
I think @Secured should be all we need. That will reference authorities, but will include a "ROLE_" prefix to make Spring Security happy. And yes, the GUI shouldn't care about roles anymore.